src/server/ua_server_ns0_gds.c

```diff
@@ -0,0 +1,1559 @@
+
+    UA_CertificateGroup *certGroup = getCertGroup(server, objectId);
+    if(!certGroup)
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+
+    /* Cannot be opened when a transaction is running */
+    if(server->transaction.state == UA_GDSTRANSACIONSTATE_PENDING)
+        return UA_STATUSCODE_BADTRANSACTIONPENDING;
+
+    UA_FileInfo *fileInfo = getFileInfo(certGroup->certificateGroupId);
+    if(!fileInfo)
+        return UA_STATUSCODE_BADINTERNALERROR;
+
+    /* Check that the list can be opened in the specified mode */
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    if(fileOpenMode == (UA_OPENFILEMODE_WRITE | UA_OPENFILEMODE_ERASEEXISTING)) {
+        if(fileInfo->openCount != 0)
+            return UA_STATUSCODE_BADNOTWRITABLE;
+    } else if(fileOpenMode == UA_OPENFILEMODE_READ) {
+        /* Nothing to check.
+         * If the list is already open for writing,
+         * the previous check for a current transaction will fail. */
+    } else {
+        return UA_STATUSCODE_BADINVALIDSTATE;
+    }
+
+    /* If the list is opened for writing, a transaction must be created */
+    if(fileOpenMode == (UA_OPENFILEMODE_WRITE | UA_OPENFILEMODE_ERASEEXISTING)) {
+        retval = UA_GDSTransaction_init(&server->transaction, server, *sessionId);
+        if(retval != UA_STATUSCODE_GOOD)
+            return retval;
+    }
+
+    if(checkSessionCallbackId == 0) {
+        retval = UA_Server_addRepeatedCallback(server, (UA_ServerCallback)checkSessionActive, NULL,
+                                       CHECKACTIVESESSIONINTERVAL, &checkSessionCallbackId);
+        if(retval != UA_STATUSCODE_GOOD)
+            return retval;
+    }
+
+    UA_TrustListDataType trustList;
+    memset(&trustList, 0, sizeof(UA_TrustListDataType));
+    trustList.specifiedLists = UA_TRUSTLISTMASKS_ALL;
+    certGroup->getTrustList(certGroup, &trustList);
+
+    UA_ByteString *encTrustList = UA_ByteString_new();
+    UA_ByteString_init(encTrustList);
+    retval = UA_encodeBinary(&trustList, &UA_TYPES[UA_TYPES_TRUSTLISTDATATYPE], encTrustList);
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_ByteString_delete(encTrustList);
+        UA_TrustListDataType_clear(&trustList);
+        return retval;
+    }
+    UA_TrustListDataType_clear(&trustList);
+
+    UA_FileContext *fileContext = (UA_FileContext*)UA_calloc(1, sizeof(UA_FileContext));
+    fileContext->file = encTrustList;
+    fileContext->sessionId = *sessionId;
+    fileContext->openFileMode = fileOpenMode;
+    fileContext->currentPos = 0;
+    retval = createFileHandleId(fileInfo, &fileContext->fileHandle);
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_ByteString_delete(fileContext->file);
+        UA_free(fileContext);
+        return retval;
+    }
+
+    fileInfo->openCount += 1;
+    UA_Variant_setScalarCopy(output, &fileContext->fileHandle, &UA_TYPES[UA_TYPES_UINT32]);
+
+    /* Updating OpenCount Varialbe in the information model */
+    UA_LOCK(&server->serviceMutex);
```

Comment:
Don't you need the lock for the entire callback?
Otherwise a parallel thread could work on the same data structures.